src/backend/distributed/executor/adaptive_executor.c

```diff
@@ -1130,13 +1128,45 @@ CleanUpSessions(DistributedExecution *execution)
 			 */
 			CloseConnection(connection);
 		}
-		else if (transactionState == REMOTE_TRANS_CLEARING_RESULTS)
+		else if (connection->connectionState == MULTI_CONNECTION_CONNECTED)
 		{
-			/* TODO: should we keep this in the state machine? or cancel? */
-			ClearResults(connection, false);
-		}
+			RemoteTransaction *transaction = &(connection->remoteTransaction);
+			RemoteTransactionState transactionState = transaction->transactionState;
 
-		/* TODO: can we be in SENT_COMMAND? */
+			if (transactionState == REMOTE_TRANS_CLEARING_RESULTS)
+			{
+				/*
+				 * We might have established the connection, and even sent BEGIN, but not
+				 * get to the point where we assigned a task to this specific connection
+				 * (because other connections in the pool already finished all the tasks).
+				 */
+				Assert(session->commandsSent == 0);
```

Comment:
It may be useful to spell out the important assumption we're making in the function comment: That we only reach `CleanUpSessions` after successfully completing all the tasks and we expect no tasks are still in progress. If we ever decide to change our logic, this function will need to be revised. (crazy example: run SELECT tasks in parallel and see which returns first)